### ISSUE TITLE
FlashAttention tutorial requires relaxed verification in advanced path (`perf_attn`)

### ISSUE THREAD

**Issue by victor-eds** [intel/intel-xpu-backend-for-triton#2098]
Comparing Triton vs XeTLA FlashAttention output in [FlashAttention](https://github.com/intel/intel-xpu-backend-for-triton/blob/7ea022d59cea9eb1660649ce8d7e9d05c13ca313/python/tutorials/06-fused-attention.forward.py) using `atol=1e-2, rtol=0` as in [upstream](https://github.com/triton-lang/triton/blob/15325e6eb1e048ba77f8c666be8b5c45cee8ab33/python/tutorials/06-fused-attention.py#L551) leads to size `1 32 16384 64` missing verification. A more relaxed `atol=1e-1` value verifies, but this might be a bit too permissive taking into account values will be less than 1 anyway (FlashAttention is a SoftMax).

In order to reproduce, add the following code to the `forward` function, right before the `return`:

```python
torch_output = torch.nn.functional.scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=0.0, is_causal=False).to(torch.float32)
torch.testing.assert_close(o, torch_output, atol=1e-2, rtol=0)
```

**Comment by Dewei-Wang-sh** [comment]
some different IR from poc, need to check they are the same.

**Comment by quintinwang5** [comment]
Actually, this is **not a problem on Triton side**.
![image](https://github.com/user-attachments/assets/8e6de982-a93a-47da-b460-4f20843a8c90)
In the computation of attention, `output = PV`, and `P` is a softmax value according the last dimension(in naive 2D case, a row). So, if we set `V` to all ones use `torch.ones`. The `output` should be all ones(or very close to 1). For `1 32 16384 64`, triton is right, but torch is not.
```
triton_output tensor([[[[1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          ...,
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000]]]],
       device='xpu:0')
===========================================================
torch tensor([[[[0.8726, 0.8726, 0.8726,  ..., 0.8726, 0.8726, 0.8726],
          [0.8726, 0.8726, 0.8726,  ..., 0.8726, 0.8726, 0.8726],
          [0.8706, 0.8706, 0.8706,  ..., 0.8706, 0.8706, 0.8706],
          ...,
         [0.8716, 0.8716, 0.8716,  ..., 0.8716, 0.8716, 0.8716],
          [0.8706, 0.8706, 0.8706,  ..., 0.8706, 0.8706, 0.8706],
          [0.8711, 0.8711, 0.8711,  ..., 0.8711, 0.8711, 0.8711]]]],
       device='xpu:0')
```
For `4 48 1024 64`, it's both OK.
```
triton_output tensor([[[[1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          ...,
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000]]]],
       device='xpu:0')
===========================================================
torch tensor([[[[1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
          ...,
          [0.9995, 0.9995, 0.9995,  ..., 0.9995, 0.9995, 0.9995],
          [0.9995, 0.9995, 0.9995,  ..., 0.9995, 0.9995, 0.9995],
```

**Comment by victor-eds** [comment]
I see for that particular corner case we are more precise (even in the second example). However, could we test this for random inputs, e.g., comparing with XeTLA and other vendors like CUDA or CPU? If this is indeed an XeTLA issue, we could report to them.

**Comment by quintinwang5** [comment]
**I've verified that Triton's result can match CUDA's using `atol=1e-2, rtol=0`. So it's clear that Pytorch gives a wrong result for this case. I'll close this issue and file a new issue to our Pytorch team.**
_Notice: If you want to verified the result, be careful with the different behaviors of `torch.manual_seed`  and `torch.randn` between `CUDA` and `XPU`. Although we choose the same seed when calling 3 `randn` for `q, k, v`, we can get the same `q`, but different `k, v`. I just save all `q, k, v`, then load the same copy to avoid this problem._

**Comment by victor-eds** [comment]
> **I've verified that Triton's result can match CUDA's using `atol=1e-2, rtol=0`. So it's clear that Pytorch gives a wrong result for this case. I'll close this issue and file a new issue to our Pytorch team.** _Notice: If you want to verified the result, be careful with the different behaviors of `torch.manual_seed` and `torch.randn` between `CUDA` and `XPU`. Although we choose the same seed when calling 3 `randn` for `q, k, v`, we can get the same `q`, but different `k, v`. I just save all `q, k, v`, then load the same copy to avoid this problem._

Thanks for the investigation! Good findings!

**Comment by quintinwang5** [comment]
Track: https://github.com/pytorch/pytorch/issues/135085

**Comment by vlad-penkin** [comment]
Let's revisit this issue, https://github.com/pytorch/pytorch/issues/135085 is closed.

**Comment by quintinwang5** [comment]
Still **OOM on 1100**, and **take too much time on 1550** (not sure it's hung or a slow execution). Used `torch=2.6.0a0+git61dc5e9` is the latest on from CI build.

**Comment by quintinwang5** [comment]
Confirmed SDPA for XPU is a feature targeted Pytorch 2.7.

**Comment by quintinwang5** [comment]
Tracking in https://github.com/pytorch/pytorch/pull/140389

**Comment by whitneywhtsang** [comment]
Both this issue and #2042 are tracking SDPA on XPU, closing this one.
